src/main/scala/zio/sqs/SqsPublisher.scala

```diff
@@ -34,4 +38,89 @@ object SqsPublisher {
       })
       ()
     }
+
+  trait Event {
+    def body: String
+    def attributes: Map[String, MessageAttributeValue]
+    def groupId: Option[String]
+    def deduplicationId: Option[String]
+  }
+
+  final case class StringEvent(body: String) extends Event {
+    override def attributes: Map[String, MessageAttributeValue] = Map.empty[String, MessageAttributeValue]
+    override def groupId: Option[String]                        = None
+    override def deduplicationId: Option[String]                = None
+  }
+
+  type ErrorOrEvent = Either[(BatchResultErrorEntry, Event), Event]
+
+  def stringStream(
+    client: SqsAsyncClient,
+    queueUrl: String,
+    settings: SqsPublisherStreamSettings = SqsPublisherStreamSettings()
+  )(
+    ms: Stream[Throwable, String]
+  ): ZStream[Clock, Throwable, ErrorOrEvent] =
+    eventStream(client, queueUrl, settings)(ms.map(StringEvent))
+
+  def eventStream(
+    client: SqsAsyncClient,
+    queueUrl: String,
+    settings: SqsPublisherStreamSettings = SqsPublisherStreamSettings()
+  )(
+    ms: Stream[Throwable, Event]
+  ): ZStream[Clock, Throwable, ErrorOrEvent] =
+    ms.aggregateAsyncWithin(Sink.collectAllN[Event](settings.batchSize.toLong), Schedule.spaced(settings.duration))
+      .map(buildSendMessageBatchRequest(queueUrl, _))
+      .mapMParUnordered(settings.parallelism)(it => runSendMessageBatchRequest(client, it._1, it._2))
+      .mapConcat(identity)
+
+  private def buildSendMessageBatchRequest(
+    queueUrl: String,
+    ms: List[Event]
+  ): (SendMessageBatchRequest, List[(Event, Int)]) = {
+    val indexedMessages = ms.zipWithIndex
+    val entries = indexedMessages.map {
+      case (m: Event, id: Int) =>
+        SendMessageBatchRequestEntry
+          .builder()
+          .id(id.toString)
+          .messageBody(m.body)
+          .messageAttributes(m.attributes.asJava)
+          .messageGroupId(m.groupId.orNull)
```

Comment:
Passing null is equivalent to not calling this? I didn't know that 😄

Reply:
yes, they are null by default in the builder